src/imputation/MoR.py

```diff
@@ -112,8 +125,148 @@ def carry_forwards(df, backdata, impute_vars):
 
+    df.loc[match_cond] = create_imp_class_col(
+        df, "200_prev", "201_prev", use_cellno=False
+    )
+
     # Drop merge related columns
-    to_drop = [column for column in df.columns if column.endswith("_prev")]
+    to_drop = [
+        column
+        for column in df.columns
+        if (column.endswith("_prev"))
+        & (re.search("(.*)_prev|.*", column).group(1) not in impute_vars)
+    ]
     to_drop += ["_merge"]
     df = df.drop(to_drop, axis=1)
     return df
+
+
+def calculate_growth_rates(current_df, prev_df, target_vars):
+    """Calculate the growth rates between previous and current data.
+
+    Args:
+        current_df (pd.DataFrame): pre-processed current data.
+        prev_df (pd.DataFrame): pre-processed backdata.
+        target_vars ([string]): target vars to impute.
+    """
+    # Only calculate links for long form responders
+    current_df = current_df.copy().loc[current_df["formtype"] == "0001", :]
+    # prev_df = prev_df.copy().loc[prev_df["formtype"] == "0001", :]
+
+    gr_df = pd.merge(
+        current_df,
+        prev_df,
+        on=["reference", "imp_class"],
+        how="inner",
+        suffixes=("", "_prev"),
+    )
+    target_vars_prev = [f"{var}_prev" for var in target_vars]
+    gr_df = (
+        gr_df[["reference", "imp_class"] + target_vars + target_vars_prev]
+        .groupby(["reference", "imp_class"])
+        .sum()
+    ).reset_index()
+    # Calculate the ratios for the relevant variables
+    for target in target_vars:
+        gr_df[f"{target}_gr"] = gr_df[target] / gr_df[f"{target}_prev"]
+
+    return gr_df
+
+
+def calculate_links(gr_df, target_vars, config):
+    """Calculate the Means of Ratios (links) for each imp_class
+
+    Args:
+        gr_df (pd.DataFrame): DataFrame of growth rates for each target variable
+        target_vars ([string]): List of target variables to use.
+        config (Dict): Confuration settings.
+    """
+    # Apply trimming and calculate means for each imp class
+    gr_df = gr_df.groupby("imp_class")
+    gr_df = gr_df.apply(group_calc_link, target_vars, config)
+
+    # Reorder columns to make QA easier
+    column_order = ["imp_class", "reference"] + list(
+        itertools.chain(
+            *[
+                (var, f"{var}_prev", f"{var}_gr", f"{var}_gr_trim", f"{var}_link")
+                for var in target_vars
+            ]
+        )
+    )
+    return gr_df[column_order].reset_index(drop=True)
+
+
+def group_calc_link(group, target_vars, config):
+    """Apply the MoR method to each group
+
+    Args:
+        group (pd.core.groupby.DataFrameGroupBy): Imputation class group
+        link_vars ([string]): List of the linked variables.
+        config (Dict): Confuration settings
+    """
+    for var in target_vars:
+        group = trim_bounds(group, f"{var}_gr", config)
+        # Create mask to not use 0s in mean calculation
+        no_zero_mask = pd.notnull(group[f"{var}_gr"]) & (group[f"{var}_gr"] != 0)
```

Comment:
A comment in here, perhaps, to explain:

 if there are no non-null, non-zero values in the trimmed group, set the link to 1

Reply:
Good shout.